### ISSUE TITLE
Dev console showing a bunch of Angular errors

### ISSUE THREAD

**Issue by ivanjuras** [soflyy/oxygen-bugs-and-features#281]
**Describe the bug**

Dev console is showing a bunch of errors when loading the Oxygen editor. Also, because of that, it seems that sometimes after I click save nothing gets saved. I only discover that after I start editing the page again -- and I have to go from scratch.

**Have you tried all the steps at https://oxygenbuilder.com/documentation/troubleshooting/troubleshooting-guide/?**

Yes

**Are you able to replicate the issue on a Sandbox install at https://oxygenbuilder.com/try?**

I can't because this shows up randomly.

**Expected behavior**

No errors, and the ability to actually save the changes.

**What actually happens**

The whole dev console gets filled out with Angularjs error messages, and sometimes, the work you save gets lost.

**Screenshots**

https://i.imgur.com/XFmVYFu.png

**Desktop (please complete the following information):**
- OS: Windows 10
 - Browser Chrome
 - Oxygen Version: 2.3 RC

**Comment by quadcom** [comment]
Disable any minification plugins and JScript deferred loading.

I've seen this happen before.

**Comment by ivanjuras** [comment]
@quadcom I'm not running any optimization plugins:

https://i.imgur.com/q6BdCLr.png

**Comment by quadcom** [comment]
I see you're running a multisite WP install. 

Check for other network activated plugins that are running that could be minifying or deferring code.

That's the only time I've seen this happen and once you turn that off, those errors go away.

**Comment by drank** [comment]
Do you by any chance have the Gravity Forms plugin active?

I see the same errors when I activate this plugin on my site (no optimization plugins, no multisite) and they disappear when I deactivate it.

If that is the case then there might be some incompatibility between Oxygen and Gravity Forms that would have to be reported.

**Comment by drank** [comment]
Here is the console output with the "debugger" URL option activated:
![image](https://user-images.githubusercontent.com/1393766/57196390-24b91480-6f65-11e9-84c1-6e7d4bcc7ef9.png)

**Comment by Gibsonion** [comment]
Since Updating to 2.3 final from 2.3 RC (I didn't think there was any changes) I've had constant errors, not just AJS one's either, in the console, divs have not been saving things, program is running more slowly, and blend modes seem to be breaking the admin bar.

My PHP error log looks like it tried to eat it self, or the dog got hold of it, all the messeages are a garbled mess, the error log look like it was having errors. Kinda glad it's not just me.

![image](https://user-images.githubusercontent.com/27298918/57196422-a6ed0d00-6f54-11e9-8c29-beed99f5b6c7.png)

![image](https://user-images.githubusercontent.com/27298918/57196500-4f9b6c80-6f55-11e9-85e5-
a18eae8b4c0b.png)


This goes on for about 10 pages. But it all seems to start exactly here:

[03-May-2019 18:03:18 UTC] WordPress database error Deadlock found when trying to get lock; try restarting transaction for query DELETE FROM wp_postmeta WHERE meta_id IN( xxxxx ) made by require('wp-blog-header.php'), 

require_once('wp-includes/template-loader.php'), include('/plugins/oxygen/component-framework/components/layouts/shortcode.php'), 

do_action('wp_footer'), WP_Hook->do_action, WP_Hook->apply_filters, SWP_Buttons_panel_Loader->floating_buttons, 

SWP_Buttons_Panel_Side->__construct, SWP_Buttons_Panel->__construct, SWP_Buttons_Panel->establish_share_data, 

SWP_Post_Cache_Loader->get_post_cache, SWP_Post_Cache->__construct, SWP_Post_Cache->rebuild_cached_data, SWP_Post_Cache->rebuild_share_counts, 

SWP_Post_Cache->cache_share_counts, delete_post_meta, delete_metadata

**Comment by drank** [comment]
OK, after some debugging I think I found the cause (and maybe a solution, at least works for me).

Here is the find:
In the Gravity Forms' js/gravityforms.min.js file there is the following code (formatted) that alters the String's prototype:
```
String.prototype.format = function () {
    var args = arguments;
    return this.replace(/{(\d+)}/g, function (match, number) {
        return typeof args[number] != 'undefined' ? args[number] : match;
    });
}; 
```
Now, in Oxygen's component-framework/angular/controllers/controller.options.js file, there is the following code, in the `$scope.isTabHasOptions` function:
```
for (var index in subtub) {
    var optionName = subtub[index];
    if ($scope.isInherited($scope.component.active.id, optionName)===false) {
        if($scope.getOption(optionName, $scope.component.active.id) != '') {
            return true;
        }
    }
}
```
Now, when Oxygen loops through the components and reaches a code block, the following happens:
In the 2nd line above, `optionName` gets assigned **a function** instead of a string - the function that has been previously declared by Gravity Forms - and that is because:
1. `subtub` is a string (not an array) and has the value of "code-php"
2. `index` gets assigned the value "format" - I guess this is because of the `for.. in...` construct applied to the String object with the altered prototype

And the above leads to `subtub[index]` evaluating to a reference to the String.prototype.format function.

Here is how I solved it (I do not pretend this to be THE solution, it just worked for me). I believe you will fix it in a much better way.
I added the following lines just above the for loop:
```
if (typeof subtub === 'string') {
    subtub = [subtub];
}
```
effectively converting `subtub` to an array if it is a string.

**Comment by drank** [comment]
Update: removing the "debugger" from the URL makes the editor to fetch iframe.min.js and my fix is gone :(

**Comment by Gibsonion** [comment]
I don't think it's got anything to do with Gravity forms as I've never used that plugin, I use Formidable

**Comment by drank** [comment]
@Gibsonion 

I think I just showed in my post above that the errors in the OP and in my screenshots, which are **Angualar and JavaScript errors**, are very much connected to Gravity Forms.

The errors that you posted are **PHP errors** and most probably are caused by other reasons, not related to the topic of this post which is "Dev console showing a bunch of Angular errors". So yes, you are right in some way - the PHP errors that you posted might not be related to Gravity Forms, but then again these errors are also not related to the OP.

Please open a separate bug report so that we do not mix topics and misguide the Devs. Thank you.

**Comment by Gibsonion** [comment]
I've also been having the Angular issues, but as the OP mentions they're inconsistent and the PHP error log was all I had to share at the time.

**Comment by KittenCodes** [comment]
@ivanjuras, Do you still see the errors in the console if you deactivate all plugins (aside from Oxygen)? If you're still seeing them, please email support@oxygenbuilder.com.

**Comment by drank** [comment]
@KittenCodes, why have you closed this?

The bug is still there in 2.4.1 (I just reproduced it) and I don't see any "fix" in the code for 3.0beta2.

I have provided enough info about the bug and the environment (and also about a potential fix) above - someone at Oxygen just needs to take it from there and implement a fix.

**Comment by KittenCodes** [comment]
@drank 

> why have you closed this?

We have an internal bug report open for console errors when using Gravity Forms with Oxygen.

**Comment by drank** [comment]
OK, got it. Thank you for the feedback!

**Comment by bomakalu** [comment]
hi was this fixed? thanks

**Comment by KittenCodes** [comment]
This is fixed in v3.2 Beta 1.
